source/common/buffer/buffer_impl.cc

```diff
@@ -409,6 +409,9 @@ uint64_t OwnedImpl::reserve(uint64_t length, RawSlice* iovecs, uint64_t num_iove
     // Check whether there are any empty slices with reservable space at the end of the buffer.
     size_t first_reservable_slice = slices_.size();
     while (first_reservable_slice > 0) {
+      // Reclaim any space that was previously reserved but never committed.
+      slices_[first_reservable_slice - 1]->clearReservation();
```

Comment:
Probably stupid question: Why do we need to keep track of reservation state at all if we have these semantics? Can't you just return allocated but not used chain space if the user asks for it again?

Reply:
Yeah, with the new semantics, we probably can stop tracking the reservation state. I'll try making that change (which should simplify the implementation) later today.